### PR TITLE
Add getRectsForRange to Paragraph

### PR DESCRIPTION
**Pull request by abarth**
We can use getRectsForRange to implement selection painting in the
framework.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

File: sky/engine/tonic/dart_class_library.h [hunk 1]
```diff
@@ -26,11 +26,27 @@ class DartClassLibrary {
...
   Dart_PersistentHandle GetClass(const DartWrapperInfo& info);
+  Dart_PersistentHandle GetClass(const std::string& library_name,
+                                 const std::string& interface_name);
 
  private:
+  Dart_PersistentHandle GetAndCacheClass(const char* library_name,
+                                         const char* interface_name,
+                                         Dart_PersistentHandle* cache_slot);
+
+  // TODO(abarth): Move this class somewhere more general.
```
Comment (chinmaygarde) [inline sky/engine/tonic/dart_class_library.h]: I like to use a variant of http://www.boost.org/doc/libs/1_35_0/doc/html/boost/hash_combine_id241013.html. So something like:
```
template <typename T>
void HashCombine(size_t& seed, T const& v) {
  seed ^= std::hash<T>()(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
}
```

Comment (abarth) [inline sky/engine/tonic/dart_class_library.h]: Yeah, this is just something basic.  I'll add the link to the comment.

File: sky/engine/core/text/Paragraph.cpp [hunk 2]
```diff
@@ -107,4 +110,34 @@ void Paragraph::paint(Canvas* canvas, const Offset& offset)
...
+  Vector<IntRect> rects;
+  for (RenderObject* object = m_renderView.get(); object; object = object->nextInPreOrder()) {
+    if (!object->isText())
+      continue;
+    RenderText* text = toRenderText(object);
+    unsigned length = text->textLength();
+    if (offset + length > start) {
+      unsigned startOffset = offset > start ? 0 : start - offset;
+      unsigned endOffset = end - offset;
+      LOG(INFO) << "startOffset=" << startOffset << " endOffset=" << endOffset;
```
Comment (chinmaygarde) [inline sky/engine/core/text/Paragraph.cpp]: Not sure you meant to leave this log in here

Comment (abarth) [inline sky/engine/core/text/Paragraph.cpp]: Oops, no.  :)

**Comment by chinmaygarde** [comment]
lgtm

File: sky/engine/core/text/Paragraph.cpp [hunk 3]
```diff
@@ -107,4 +110,33 @@ void Paragraph::paint(Canvas* canvas, const Offset& offset)
...
 }
 
+std::vector<Rect> Paragraph::getRectsForRange(unsigned start, unsigned end) {
+  if (end <= start || start == end)
+    return std::vector<Rect>();
+
+  unsigned offset = 0;
+  Vector<IntRect> rects;
+  for (RenderObject* object = m_renderView.get(); object; object = object->nextInPreOrder()) {
+    if (!object->isText())
```
Comment (Hixie) [inline sky/engine/core/text/Paragraph.cpp]: we probably want to show inline replaced boxes (e.g. images) as selected too (once we have a way to have "replaced boxes")

Comment (Hixie) [inline sky/engine/core/text/Paragraph.cpp]: (obviously not necessarily in this patch)

**Pull request merged**